compiler/src/iree/compiler/Codegen/TransformStrategies/GPU/AbstractGemmLikeStrategy.cpp

```diff
@@ -85,17 +77,87 @@ void AbstractGemmLikeStrategy::initDefaultValues() {
   pipelineDepth = clPipelineDepth;
   peelPipelineEpilogue = clPeelPipelineEpilogue;
 
-  if (!clBlockTileSizes.isDefaultAssigned() ||
-      !clNumThreads.isDefaultAssigned() || !clNumWarps.isDefaultAssigned() ||
-      reductionTileSize != clReductionTileSize.getDefault().getValue() ||
-      useAsyncCopies != clUseAsyncCopies.getDefault().getValue() ||
-      useMmaSync != clUseMmaSync.getDefault().getValue() ||
-      useWmma != clUseWmma.getDefault().getValue() ||
-      useFma != clUseFma.getDefault().getValue() ||
-      pipelineDepth != clPipelineDepth.getDefault().getValue() ||
-      peelPipelineEpilogue != clPeelPipelineEpilogue.getDefault().getValue()) {
+  /// cliOptionsSpecified is used to override hard-coded well known good
+  /// defaults when set.
+  if (clBlockTileSizes.getNumOccurrences() ||
+      clNumThreads.getNumOccurrences() || clNumWarps.getNumOccurrences() ||
+      clReductionTileSize.getNumOccurrences() ||
+      clUseAsyncCopies.getNumOccurrences() ||
+      clUseMmaSync.getNumOccurrences() || clUseWmma.getNumOccurrences() ||
+      clUseFma.getNumOccurrences() || clPipelineDepth.getNumOccurrences() ||
+      clPeelPipelineEpilogue.getNumOccurrences()) {
     cliOptionsSpecified = true;
   }
+
+  /// Default configuration based on hardware properties and problem bit widths.
+  if (clBlockTileSizes.getNumOccurrences()) {
+    blockTileSizes =
+        SmallVector<int64_t>(clBlockTileSizes.begin(), clBlockTileSizes.end());
+  } else {
+    blockTileSizes = SmallVector<int64_t>{128, 128, 1};
+  }
+
+  if (clNumThreads.getNumOccurrences()) {
+    numThreads = SmallVector<int64_t>(clNumThreads.begin(), clNumThreads.end());
+  } else {
+    // Infer from warp counts if present.
+    if (clNumWarps.getNumOccurrences()) {
+      numThreads = SmallVector<int64_t>(clNumWarps.begin(), clNumWarps.end());
+      numThreads[0] *= gpuModel.subgroupSize;
+    } else {
+      numThreads = SmallVector<int64_t>{64, 2, 1};
+    }
+  }
+  if (clNumWarps.getNumOccurrences()) {
+    numWarps = SmallVector<int64_t>(clNumWarps.begin(), clNumWarps.end());
+  } else {
+    numWarps = numThreads;
+    numWarps[0] = mlir::ceilDiv(numWarps[0], gpuModel.subgroupSize);
+  }
+  /// TODO: Restrict based on hardware support.
```

Comment:
The GPUModel doesn't hold enough information to do that here, does it?

Reply:
Actually I forgot to remove this comment. My understanding is that all targets that support mma.sync also support async copies so I'm just using that for now.